pysb/bng.py

```diff
@@ -924,6 +924,47 @@ def _is_bool_expr(e):
     return isinstance(e, Boolean) and not isinstance(e, sympy.AtomicExpr)
 
 
+# BNG's expression syntax supports multiplication with a boolean, e.g.
+# `2 * (x > 0)`, but sympy does not. Our overall strategy is to convert
+# multiplication with a boolean expression into a BNG-equivalent Piecewise,
+# leaving other multiplications alone. First, we use a transformation function
+# during parsing to temporarily replace any '*' operators with a fake 'bngl_mul'
+# function call. Then, after parsing and evaluation, we perform a subexpression
+# replacement on the 'bngl_mul' calls that converts Boolean args to a Piecewise
+# and emits a normal Mul. This can't be done entirely in a parser transformation
+# because sympy hasn't yet parsed the expression to determine the AST structure
+# that lets us identify boolean subexpressions, and it can't be done entirely
+# post-parsing because multiplication of booleans raises an immediate error
+# during the evaluation step in parse_expr. This function is the parser
+# transformation.
+def _convert_mul(tokens, local_dict, global_dict):
+    res1 = sympy_parser._group_parentheses(_convert_mul)(tokens, local_dict, global_dict)
```

Comment:
this recursively applies this function to contents of parentheses groups, similar to traversing the AST, right?

Reply:
Yes. Sympy already provides this implementation to handle similar use cases like converting `a = b` to `Eq(a, b)`.

Reply:
I just realized that `convert_equals_signs` transform is actually pretty sketchy and doesn't even work properly when the `=` expression is in certain contexts, so it was kind of a bad example to learn from.